<a href="https://colab.research.google.com/github/deepakpracheta/Supervised-ML-Regression-Appliances-Energy-Prediction/blob/main/Supervised_ML_Appliances_Energy_Prediction.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


Data-driven prediction of energy use of appliances


The data set is at 10 min for about 4.5 months. The house temperature and humidity conditions
were monitored with a ZigBee wireless sensor network. Each wireless node transmitted the
temperature and humidity conditions around 3.3 min. Then, the wireless data was averaged for
10 minutes periods. The energy data was logged every 10 minutes with m-bus energy meters.
Weather from the nearest airport weather station (Chievres Airport, Belgium) was downloaded
from a public data set from Reliable Prognosis (rp5.ru) and merged together with the
experimental data sets using the date and time column. Two random variables have been
included in the data set for testing the regression models and to filter out non-predictive attributes
(parameters).


In [2]:
#installing package
!pip install scikit-optimize

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     |████████████████████████████████| 100 kB 3.7 MB/s 


# Importing Relevant Libraries

In [3]:
# Importing the libraries
import numpy as np
import pandas as pd
from numpy import math
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from sklearn.ensemble import RandomForestRegressor
from sklearn.ensemble import GradientBoostingRegressor
from xgboost import XGBRegressor
from lightgbm import LGBMRegressor
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_absolute_error
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
from sklearn.model_selection import RandomizedSearchCV

In [4]:
#importing dataset
path='/content/drive/MyDrive/Capstone Projects/Supervised ML-Regression Appliances Energy Prediction/data_application_energy.csv'
df=pd.read_csv(path)


In [5]:
df.head().T

0                    1                    2  \
date         2016-01-11 17:00:00  2016-01-11 17:10:00  2016-01-11 17:20:00   
Appliances                    60                   60                   50   
lights                        30                   30                   30   
T1                         19.89                19.89                19.89   
RH_1                   47.596667            46.693333                 46.3   
T2                          19.2                 19.2                 19.2   
RH_2                       44.79              44.7225            44.626667   
T3                         19.79                19.79                19.79   
RH_3                       44.73                44.79            44.933333   
T4                          19.0                 19.0            18.926667   
RH_4                   45.566667              45.9925                45.89   
T5                     17.166667            17.166667            17.166667   
RH_5                        55.2                 55.2                55.09   
T6                      7.026667             6.833333                 6.56   
RH_6                   84.256667            84.063333            83.156667   
T7                          17.2                 17.2                 17.2   
RH_7                   41.626667                41.56            41.433333   
T8                          18.2                 18.2                 18.2   
RH_8                        48.9            48.863333                48.73   
T9                     17.033333            17.066667                 17.0   
RH_9                       45.53                45.56                 45.5   
T_out                        6.6             6.483333             6.366667   
Press_mm_hg                733.5                733.6                733.7   
RH_out                      92.0                 92.0                 92.0   
Windspeed                    7.0             6.666667             6.333333   
Visibility                  63.0            59.166667            55.333333   
Tdewpoint                    5.3                  5.2                  5.1   
rv1                    13.275433            18.606195            28.642668   
rv2                    13.275433            18.606195            28.642668   

                               3                    4  
date         2016-01-11 17:30:00  2016-01-11 17:40:00  
Appliances                    50                   60  
lights                        40                   40  
T1                         19.89                19.89  
RH_1                   46.066667            46.333333  
T2                          19.2                 19.2  
RH_2                       44.59                44.53  
T3                         19.79                19.79  
RH_3                        45.0                 45.0  
T4                         18.89                18.89  
RH_4                   45.723333                45.53  
T5                     17.166667                 17.2  
RH_5                       55.09                55.09  
T6                      6.433333             6.366667  
RH_6                   83.423333            84.893333  
T7                     17.133333                 17.2  
RH_7                       41.29                41.23  
T8                          18.1                 18.1  
RH_8                       48.59                48.59  
T9                          17.0                 17.0  
RH_9                        45.4                 45.4  
T_out                       6.25             6.133333  
Press_mm_hg                733.8                733.9  
RH_out                      92.0                 92.0  
Windspeed                    6.0             5.666667  
Visibility                  51.5            47.666667  
Tdewpoint                    5.0                  4.9  
rv1                    45.410389            10.084097  
rv2                    45.410389            10.084097

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         19735 non-null  object 
 1   Appliances   19735 non-null  int64  
 2   lights       19735 non-null  int64  
 3   T1           19735 non-null  float64
 4   RH_1         19735 non-null  float64
 5   T2           19735 non-null  float64
 6   RH_2         19735 non-null  float64
 7   T3           19735 non-null  float64
 8   RH_3         19735 non-null  float64
 9   T4           19735 non-null  float64
 10  RH_4         19735 non-null  float64
 11  T5           19735 non-null  float64
 12  RH_5         19735 non-null  float64
 13  T6           19735 non-null  float64
 14  RH_6         19735 non-null  float64
 15  T7           19735 non-null  float64
 16  RH_7         19735 non-null  float64
 17  T8           19735 non-null  float64
 18  RH_8         19735 non-null  float64
 19  T9  

In our dataset all columns are numerical type except date column which is of date-time format

In [ ]:
df.shape

(19735, 29)

In our dataset there are 29 columns and 19735 rows.

In [ ]:
df.describe()

Appliances        lights            T1          RH_1            T2  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      97.694958      3.801875     21.686571     40.259739     20.341219   
std      102.524891      7.935988      1.606066      3.979299      2.192974   
min       10.000000      0.000000     16.790000     27.023333     16.100000   
25%       50.000000      0.000000     20.760000     37.333333     18.790000   
50%       60.000000      0.000000     21.600000     39.656667     20.000000   
75%      100.000000      0.000000     22.600000     43.066667     21.500000   
max     1080.000000     70.000000     26.260000     63.360000     29.856667   

               RH_2            T3          RH_3            T4          RH_4  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      40.420420     22.267611     39.242500     20.855335     39.026904   
std        4.069813      2.006111      3.254576      2.042884      4.341321   
min       20.463333     17.200000     28.766667     15.100000     27.660000   
25%       37.900000     20.790000     36.900000     19.530000     35.530000   
50%       40.500000     22.100000     38.530000     20.666667     38.400000   
75%       43.260000     23.290000     41.760000     22.100000     42.156667   
max       56.026667     29.236000     50.163333     26.200000     51.090000   

       ...            T9          RH_9         T_out   Press_mm_hg  \
count  ...  19735.000000  19735.000000  19735.000000  19735.000000   
mean   ...     19.485828     41.552401      7.411665    755.522602   
std    ...      2.014712      4.151497      5.317409      7.399441   
min    ...     14.890000     29.166667     -5.000000    729.300000   
25%    ...     18.000000     38.500000      3.666667    750.933333   
50%    ...     19.390000     40.900000      6.916667    756.100000   
75%    ...     20.600000     44.338095     10.408333    760.933333   
max    ...     24.500000     53.326667     26.100000    772.300000   

             RH_out     Windspeed    Visibility     Tdewpoint           rv1  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      79.750418      4.039752     38.330834      3.760707     24.988033   
std       14.901088      2.451221     11.794719      4.194648     14.496634   
min       24.000000      0.000000      1.000000     -6.600000      0.005322   
25%       70.333333      2.000000     29.000000      0.900000     12.497889   
50%       83.666667      3.666667     40.000000      3.433333     24.897653   
75%       91.666667      5.500000     40.000000      6.566667     37.583769   
max      100.000000     14.000000     66.000000     15.500000     49.996530   

                rv2  
count  19735.000000  
mean      24.988033  
std       14.496634  
min        0.005322  
25%       12.497889  
50%       24.897653  
75%       37.583769  
max       49.996530  

[8 rows x 28 columns]

There is none catergorical column in the dataset.
There is no null or missing values.
Max Energy consumption is 1080 and min is 10.
Light Columns has majority of values as 0.


In [ ]:
df.isnull().sum()

date           0
Appliances     0
lights         0
T1             0
RH_1           0
T2             0
RH_2           0
T3             0
RH_3           0
T4             0
RH_4           0
T5             0
RH_5           0
T6             0
RH_6           0
T7             0
RH_7           0
T8             0
RH_8           0
T9             0
RH_9           0
T_out          0
Press_mm_hg    0
RH_out         0
Windspeed      0
Visibility     0
Tdewpoint      0
rv1            0
rv2            0
dtype: int64

In [ ]:
df.corr()

Appliances    lights        T1      RH_1        T2      RH_2  \
Appliances     1.000000  0.197278  0.055447  0.086031  0.120073 -0.060465   
lights         0.197278  1.000000 -0.023528  0.106968 -0.005622  0.050985   
T1             0.055447 -0.023528  1.000000  0.164006  0.836834 -0.002509   
RH_1           0.086031  0.106968  0.164006  1.000000  0.269839  0.797535   
T2             0.120073 -0.005622  0.836834  0.269839  1.000000 -0.165610   
RH_2          -0.060465  0.050985 -0.002509  0.797535 -0.165610  1.000000   
T3             0.085060 -0.097393  0.892402  0.253230  0.735245  0.137319   
RH_3           0.036292  0.131161 -0.028550  0.844677  0.121497  0.678326   
T4             0.040281 -0.008859  0.877001  0.106180  0.762066 -0.047304   
RH_4           0.016965  0.114936  0.097861  0.880359  0.231563  0.721435   
T5             0.019760 -0.078745  0.885247  0.205797  0.720550  0.110409   
RH_5           0.006955  0.141233 -0.014782  0.303258  0.029595  0.250271   
T6             0.117638 -0.079029  0.654769  0.316141  0.801186 -0.009670   
RH_6          -0.083178  0.153756 -0.615045  0.245126 -0.580372  0.389933   
T7             0.025801 -0.135347  0.838705  0.021397  0.663660 -0.051422   
RH_7          -0.055642  0.035069  0.135182  0.801122  0.229212  0.690584   
T8             0.039572 -0.071458  0.825413 -0.030053  0.578191 -0.041023   
RH_8          -0.094039  0.012915 -0.006441  0.736196  0.068534  0.679777   
T9             0.010010 -0.157592  0.844777  0.115263  0.675535  0.054544   
RH_9          -0.051462 -0.008766  0.071756  0.764001  0.157346  0.676467   
T_out          0.099155 -0.074424  0.682846  0.340767  0.792255  0.033674   
Press_mm_hg   -0.034885 -0.010576 -0.150574 -0.293957 -0.133028 -0.255646   
RH_out        -0.152282  0.068543 -0.345481  0.274126 -0.505291  0.584911   
Windspeed      0.087122  0.060281 -0.087654  0.204932  0.052495  0.069190   
Visibility     0.000230  0.020038 -0.076210 -0.021057 -0.069721 -0.005368   
Tdewpoint      0.015353 -0.036322  0.571309  0.639106  0.582602  0.499152   
rv1           -0.011145  0.000521 -0.006203 -0.000699 -0.011087  0.006275   
rv2           -0.011145  0.000521 -0.006203 -0.000699 -0.011087  0.006275   

                   T3      RH_3        T4      RH_4  ...        T9      RH_9  \
Appliances   0.085060  0.036292  0.040281  0.016965  ...  0.010010 -0.051462   
lights      -0.097393  0.131161 -0.008859  0.114936  ... -0.157592 -0.008766   
T1           0.892402 -0.028550  0.877001  0.097861  ...  0.844777  0.071756   
RH_1         0.253230  0.844677  0.106180  0.880359  ...  0.115263  0.764001   
T2           0.735245  0.121497  0.762066  0.231563  ...  0.675535  0.157346   
RH_2         0.137319  0.678326 -0.047304  0.721435  ...  0.054544  0.676467   
T3           1.000000 -0.011234  0.852778  0.122737  ...  0.901324  0.134602   
RH_3        -0.011234  1.000000 -0.140457  0.898978  ... -0.195270  0.833538   
T4           0.852778 -0.140457  1.000000 -0.048650  ...  0.889439 -0.025549   
RH_4         0.122737  0.898978 -0.048650  1.000000  ... -0.044518  0.856591   
T5           0.888169 -0.050062  0.871813  0.091812  ...  0.911055  0.072308   
RH_5        -0.066355  0.375422 -0.076489  0.352591  ... -0.138509  0.272197   
T6           0.686882  0.076833  0.652350  0.259047  ...  0.667177  0.184424   
RH_6        -0.647672  0.514912 -0.703149  0.392178  ... -0.738940  0.391943   
T7           0.847374 -0.250090  0.877763 -0.131204  ...  0.944776 -0.077690   
RH_7         0.172624  0.832685  0.043527  0.894301  ...  0.028055  0.858686   
T8           0.795283 -0.283228  0.796256 -0.167066  ...  0.869338 -0.156820   
RH_8         0.044427  0.828822 -0.095192  0.847259  ... -0.113014  0.855812   
T9           0.901324 -0.195270  0.889439 -0.044518  ...  1.000000 -0.008683   
RH_9         0.134602  0.833538 -0.025549  0.856591  ... -0.008683  1.000000   
T_out        0.699417  0.118207  0.663478  0.293289  ...  0.668220  0.223270   
Press_mm_hg -0.1899

## Columns Information



1.  Date time year-month-day hour:minute:second
2.  Appliances, energy use in Wh
3.  Lights, energy use of light fixtures in the house in Wh
4.  T1, Temperature in kitchen area, in Celsius
5.  RH_1, Humidity in kitchen area, in %
6.  T2, Temperature in living room area, in Celsius
7.  RH_2, Humidity in living room area, in %
8.  T3, Temperature in laundry room area
9.  RH_3, Humidity in laundry room area, in %
10. T4, Temperature in office room, in Celsius
11. RH_4, Humidity in office room, in %
12. T5, Temperature in bathroom, in Celsius
13. RH_5, Humidity in bathroom, in %
14. T6, Temperature outside the building (north side), in Celsius
15. RH_6, Humidity outside the building (north side), in %
16. T7, Temperature in ironing room , in Celsius
17. RH_7, Humidity in ironing room, in %
18. T8, Temperature in teenager room 2, in Celsius
19. RH_8, Humidity in teenager room 2, in %
20. T9, Temperature in parents room, in Celsius
21. RH_9, Humidity in parents room, in %
22. To, Temperature outside (from Chievres weather station), in Celsius
23. Pressure (from Chievres weather station), in mm Hg
24. RH_out, Humidity outside (from Chievres weather station), in %
25. Wind speed (from Chievres weather station), in m/s
26. Visibility (from Chievres weather station), in km
27. Tdewpoint (from Chievres weather station), °C
28. rv1, Random variable 1, nondimensional
29. rv2, Random variable 2, nondimensional

In [7]:
df.columns

Index(['date', 'Appliances', 'lights', 'T1', 'RH_1', 'T2', 'RH_2', 'T3',
       'RH_3', 'T4', 'RH_4', 'T5', 'RH_5', 'T6', 'RH_6', 'T7', 'RH_7', 'T8',
       'RH_8', 'T9', 'RH_9', 'T_out', 'Press_mm_hg', 'RH_out', 'Windspeed',
       'Visibility', 'Tdewpoint', 'rv1', 'rv2'],
      dtype='object')

In [ ]:
missing_cols = df.isnull().mean()

In [ ]:
missing_cols.sort_values(ascending=False)

date           0.0
T7             0.0
rv1            0.0
Tdewpoint      0.0
Visibility     0.0
Windspeed      0.0
RH_out         0.0
Press_mm_hg    0.0
T_out          0.0
RH_9           0.0
T9             0.0
RH_8           0.0
T8             0.0
RH_7           0.0
RH_6           0.0
Appliances     0.0
T6             0.0
RH_5           0.0
T5             0.0
RH_4           0.0
T4             0.0
RH_3           0.0
T3             0.0
RH_2           0.0
T2             0.0
RH_1           0.0
T1             0.0
lights         0.0
rv2            0.0
dtype: float64